File: Day289_2025/Q289_GFG.java
class Q289_GFG {
    // simple Node definition so the type is resolved
    static class Node {
        int data;
        Node left, right;
        Node(int data) { this.data = data; }
    }

    Node removekeys(Node root, int l, int r) {
        if (root == null) return null;

        // If current node is smaller than L, then its left subtree is useless
        if (root.data < l)
            return removekeys(root.right, l, r);

        // If current node is greater than R, then its right subtree is useless
        if (root.data > r)
            return removekeys(root.left, l, r);

        // Otherwise, fix the left and right subtrees
        root.left = removekeys(root.left, l, r);
        root.right = removekeys(root.right, l, r);

        return root;
    }
}

File: Day289_2025/Q289_GFG.py
class Solution:
    def removekeys(self, root, l, r):
        if not root:
            return None
        
        # If current node is smaller than L, skip to right subtree
        if root.data < l:
            return self.removekeys(root.right, l, r)
        
        # If current node is greater than R, skip to left subtree
        if root.data > r:
            return self.removekeys(root.left, l, r)
        
        # Otherwise, fix both subtrees
        root.left = self.removekeys(root.left, l, r)
        root.right = self.removekeys(root.right, l, r)
        
        return root
